baseline/baseline.py

```diff
@@ -0,0 +1,145 @@
+        num += 1; print(num)
+
+        query = item['corrected_question']
+        # text_ents, standard_ents, standard_ent_uries, confs, types = EntityLinking(query, 'more')
+        standard_ents, text_ents = Entity_Link_Falcon(query)
+
+        for i in range(len(standard_ents)):
+            standard_ent = standard_ents[i]
+            predicate_uris = GetPredicateList(standard_ent, template_id=item['sparql_template_id'])
+            count += 1
+
+            for predicate_uri in predicate_uris:
+                predicate = predicate_uri.split('/')[-1]
+                if predicate not in idf:
+                    idf[predicate] = 1
+                else:
+                    idf[predicate] += 1
+
+    for k, v in idf.items():
+        idf[k] = math.log2(float(count)/v)
+    with open('../data/idf.pkl', 'wb') as f:
+        pickle.dump(idf, f)
+    
+    print('get idf done!')
+
+    return idf
+
+
+def method1(idf):
+    simple_queries = get_simple_query(paths.lcquad_test)
+
+    total_res = []
+    count = 0
+
+    for item in simple_queries:
+
+        count += 1; print(count)
+
+        query = item['corrected_question']; query_words = word_tokenize(query)
+        query_id = item['_id']
+
+        # text_ents, standard_ents, standard_ent_uries, confs, types = EntityLinking(query, 'more')
+        standard_ents, text_ents = Entity_Link_Falcon(query)
+
+        final_entity = ''; final_predicate = ''; final_score = float('-inf')
+        for i in range(len(standard_ents)):
+            text_ent = text_ents[i]; standard_ent = standard_ents[i]
+            sentence_emb = []
+
+            tmp = []
+            for word in query_words:
+                if word not in text_ent:
+                    tmp.append(word)
+            
+            gram_2 = get_ngram(tmp, 2)
+            for gram in gram_2:
+                # if not g.emb(word)[0]:
+                #     sentence_emb.append(np.random.uniform(-0.01, 0.01, size=(1, 300))[0])
+                # else:
+                #     sentence_emb.append(np.array(g.emb(word)))
+                emb1 = fasttext[gram[0]].numpy(); emb2 = fasttext[gram[1]].numpy()
+                sentence_emb.append(emb1+emb2)
+            sentence_emb = np.array(sentence_emb)
+            
+            predicate_uris = GetPredicateList(standard_ent, template_id=item['sparql_template_id'])
+
+            ans_predicate = ''; max_score = float('-inf')
+            for predicate_uri in predicate_uris:
+                predicate = predicate_uri.split('/')[-1]
+                if predicate in idf:
+                    idf_score = idf[predicate]
+                else:
+                    idf_score = 2.0
+
+                # if not g.emb(predicate)[0]:
+                #     predicate_emb = np.array([np.random.uniform(-0.01, 0.01, size=(1, 300))[0]])
+                # else:
+                #     predicate_emb = np.array([g.emb(predicate)])
+                predicate_emb = np.array([fasttext[predicate].numpy()])
```

Comment:
Please do word tokenization here. Some predicates are multi word. Calculate embedding for each word and then take an average.

`Eg: <http://dbpedia.org/ontology/routeEnd> => 'route end' which is the 3rd question in the results`